File: Claude/Python/2025/Day2_Solution_2025.py
with open("Day2_Input.txt") as f:
    data = f.read().strip()

ranges = []
for part in data.split(','):
    part = part.strip()
    if not part:
        continue
    a, b = part.split('-')
    ranges.append((int(a), int(b)))

# Generate all "doubled" numbers: digits repeated twice, no leading zeros
# prefix length 1..5 covers up to 10-digit numbers (max in input ~9.5e9)
doubled = []
for length in range(1, 6):
    start = 10 ** (length - 1)
    end = 10 ** length
    for prefix in range(start, end):
        s = str(prefix)
        n = int(s + s)
        doubled.append(n)

doubled.sort()

total = 0
for a, b in ranges:
    for n in doubled:
        if n > b:
            break
        if n >= a:
            total += n

print(total)

File: Claude/Rust/2025/Day2_Solution_2025.rs
use std::time::Instant;
use std::fs;

fn main() {
    let start = Instant::now();
    let content = fs::read_to_string("Day2_Input.txt").expect("Could not read input file");
    let content = content.trim();

    let ranges: Vec<(u64, u64)> = content
        .split(',')
        .filter(|s| !s.is_empty())
        .map(|range| {
            let (lo, hi) = range.split_once('-').expect("Invalid range format");
            (lo.parse().unwrap(), hi.parse().unwrap())
        })
        .collect();

    let mut total: u64 = 0;

    // Generate all "doubled" numbers: a number whose decimal form is some
    // sequence X concatenated with itself (e.g. 11, 6464, 123123).
    // No leading zeros means X itself has no leading zero, so X >= 10^(half_len-1).
    // We only need half_len 1..=5 since max input value fits in 10 digits.
    for half_len in 1u32..=5 {
        let half_start = 10u64.pow(half_len - 1);
        let half_end = 10u64.pow(half_len);
        let multiplier = half_end; // 10^half_len shifts left by half_len digits

        for half in half_start..half_end {
            let doubled = half * multiplier + half;
            for &(lo, hi) in &ranges {
                if doubled >= lo && doubled <= hi {
                    total += doubled;
                    break;
                }
            }
        }
    }

    println!("{}", total);

    println!("Runtime: {:.6} seconds", start.elapsed().as_secs_f64());
}
